src/timelock.cairo

```diff
@@ -125,26 +177,47 @@ mod Timelock {
 
         fn cancel(ref self: ContractState, id: felt252) {
             self.check_owner();
-            assert(self.execution_started.read(id).is_non_zero(), 'DOES_NOT_EXIST');
-            assert(self.executed.read(id).is_zero(), 'ALREADY_EXECUTED');
-
-            self.execution_started.write(id, 0);
+            let execution_state = self.execution_state.read(id);
+            assert(execution_state.started.is_non_zero(), 'DOES_NOT_EXIST');
+            assert(execution_state.executed.is_zero(), 'ALREADY_EXECUTED');
+
+            self
+                .execution_state
+                .write(
+                    id,
+                    ExecutionState {
+                        started: 0,
+                        executed: execution_state.executed,
+                        canceled: execution_state.canceled
+                    }
+                );
 
             self.emit(Canceled { id, });
         }
 
         fn execute(ref self: ContractState, mut calls: Span<Call>) -> Array<Span<felt252>> {
             let id = to_id(calls);
 
-            assert(self.executed.read(id).is_zero(), 'ALREADY_EXECUTED');
+            let execution_state = self.execution_state.read(id);
 
-            let (earliest, latest) = self.get_execution_window(id);
+            assert(execution_state.executed.is_zero(), 'ALREADY_EXECUTED');
+
+            let execution_window = self.get_execution_window(id);
```

Comment:
this causes a second read of the same slot

not super costly, but wasteful
better to revert if already executed or canceled from this function, since there is no execution window